### DISCUSSION TITLE
CVC5 and parallelism

### DISCUSSION THREAD

**Discussion by daniel-raffler** [cvc5/cvc5#10265] [General]
Hello everybody, 
we've been using CVC5 in [JavaSMT](https://github.com/sosy-lab/java-smt) as one of our solver backends. Recently we've noticed some issues when the solver is used in a parallel on several threads. I've already reported one of these bugs [here](https://github.com/cvc5/cvc5/issues/10256), however, there are several more that have been found by our testsuite. I rewrote the test to use CVC5 directly and the C++ versions can now be found [here](https://github.com/daniel-raffler/cvc5/commit/34d7bbbfcadf26343d97e60061f5d7d91814cf20). The issue seems to be that CVC5 uses one NodeManager per thread and doesn't allow solver structures to change from one thread to another. I was able to fix the issue by simply removing the thread_local modifier ([here](https://github.com/daniel-raffler/cvc5/commit/f4fe8ef9eb92de6b5f51c4b50234f71dda3feb80)), but this introduces a host of new problems.

Is parallel support still being worked on? We don't necessarily need parallel solving, but it would be nice to at least be able to move solver instances between threads. Maybe this could be accomplished by exposing the NodeManager in the API?

At the very least it would be useful to get some more descriptive error messages. We've seen error messages ranging from "Number is not an integer value: 4" to  "Fatal failure within static cvc5::internal::TypeNode cvc5::internal::expr::TypeChecker::computeType(cvc5::internal::NodeManager*, cvc5::internal::TNode, bool, std::ostream*)" when the problem is really that the wrong thread is used. The issue seems to be that CVC5 doesn't immediately fail if the wrong thread is used, but instead returns "null instances" if the needed object can't be found in the NodeManager. This then leads to a variety of rather confusing error messages down the line.

Thanks in advance,
Daniel

**Comment by mpreiner** [comment]
We currently do not fully support solver instances running in different threads due to the node manager being thread_local. As soon as you remove this keyword all bets are off since the node manager itself is not thread-safe. We are working on a solution for this, but this will take some time since this will require substantial changes to the internals of cvc5 and its public API. Just to make sure: We are aware of the issue and we are working on a solution.

**Comment by daniel-raffler** [comment]
Hallo Mathias,
Thanks for looking into this! I understand that the redesign may take some time and I'll simply disable the affected tests for now.
